plugins/modules/ndo_mac_sec_policy.py

```diff
@@ -0,0 +1,423 @@
+import copy
+
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        dict(
+            template=dict(type="str", required=True),
+            mac_sec_policy=dict(type="str", aliases=["name"]),
+            mac_sec_policy_uuid=dict(type="str", aliases=["uuid"]),
+            description=dict(type="str"),
+            admin_state=dict(type="str", choices=["enabled", "disabled"]),
+            interface_type=dict(type="str", choices=["fabric", "access"], default="fabric"),
+            cipher_suite=dict(type="str", choices=["128_gcm_aes", "128_gcm_aes_xpn", "256_gcm_aes", "256_gcm_aes_xpn"]),
+            window_size=dict(type="int"),
+            security_policy=dict(type="str", choices=["should_secure", "must_secure"]),
+            sak_expiry_time=dict(type="int"),
+            confidentiality_offset=dict(type="int", choices=[0, 30, 50]),
+            key_server_priority=dict(type="int"),
+            mac_sec_key=dict(
+                type="list",
+                elements="dict",
+                options=dict(
+                    key_name=dict(type="str", required=True),
+                    psk=dict(type="str", required=True, no_log=True),
+                    start_time=dict(type="str"),
+                    end_time=dict(type="str"),
+                ),
+                no_log=False,
+            ),
+            state=dict(type="str", choices=["absent", "query", "present"], default="query"),
+        )
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ["state", "present", ["mac_sec_policy"]],
+            ["state", "absent", ["mac_sec_policy"]],
+        ],
+    )
+
+    mso = MSOModule(module)
+
+    template = module.params.get("template")
+    mac_sec_policy = module.params.get("mac_sec_policy")
+    mac_sec_policy_uuid = module.params.get("mac_sec_policy_uuid")
+    description = module.params.get("description")
+    admin_state = module.params.get("admin_state")
+    interface_type = module.params.get("interface_type")
+    cipher_suite = module.params.get("cipher_suite")
+    window_size = module.params.get("window_size")
+    security_policy = module.params.get("security_policy")
+    sak_expiry_time = module.params.get("sak_expiry_time")
+    confidentiality_offset = module.params.get("confidentiality_offset")
+    key_server_priority = module.params.get("key_server_priority")
+    mac_sec_keys = module.params.get("mac_sec_key")
+    state = module.params.get("state")
+
+    ops = []
+    match = None
+
+    mso_template = MSOTemplate(mso, "fabric_policy", template)
+    mso_template.validate_template("fabricPolicy")
+
+    path = "/fabricPolicyTemplate/template/macsecPolicies"
+
+    existing_mac_sec_policies = mso_template.template.get("fabricPolicyTemplate", {}).get("template", {}).get("macsecPolicies", [])
+    if mac_sec_policy or mac_sec_policy_uuid:
```

Comment:
This changes behaviour from other ndo template modules, due to query now possible on uuid. If we want this enhancement we should directly change this in the other ndo modules, ndo_dhcp_option_policy, ndo_dhcp_relay_policy, ndo_l3_domain, ndo_physical_domain, ndo_route_map_policy_multicast, ndo_synce_interface_policy, and ndo_vlan_pool.

Personally I do not see the value of uuid query, due to that user are not exposed to this internal uuid and need to specifically query for it, which would be the name based query thus adding a duplicate task.

Reply:
Created new issue for this. https://github.com/CiscoDevNet/ansible-mso/issues/537